PetAdoptions/pethistory/template.yaml

```diff
@@ -0,0 +1,222 @@
+          LAMBDA_APPLICATION_SIGNALS_REMOTE_ENVIRONMENT: lambda:default
+          OTEL_METRICS_EXPORTER: none
+          OTEL_LOGS_EXPORTER: none
+
+      # AWS Distro for OpenTelemetry Layer for auto-instrumentation
+      # Using the correct AWS-managed OTEL layer for us-west-2
+      Layers:
+        - arn:aws:lambda:us-west-2:615299751070:layer:AWSOpenTelemetryDistroJs:8
+
+      # SQS Event Source
+      Events:
+        SQSEvent:
+          Type: SQS
+          Properties:
+            Queue: !Ref SQSQueueArn
+            BatchSize: 10
+            MaximumBatchingWindowInSeconds: 5
+            FunctionResponseTypes:
+              - ReportBatchItemFailures
+
+      # IAM Policies
+      Policies:
+        - AWSLambdaBasicExecutionRole
+        - AWSLambdaVPCAccessExecutionRole
+
+        # Secrets Manager access
+        - Version: "2012-10-17"
+          Statement:
+            - Effect: Allow
+              Action:
+                - secretsmanager:GetSecretValue
+              Resource: !Ref RDSSecretArn
+
+        # RDS access (if using RDS Proxy or IAM database authentication)
+        - Version: "2012-10-17"
+          Statement:
+            - Effect: Allow
+              Action:
+                - rds-db:connect
+              Resource: !Sub "arn:aws:rds-db:${AWS::Region}:${AWS::AccountId}:dbuser:*/lambda-user"
+
+        # X-Ray and CloudWatch Application Signals permissions
+        - Version: "2012-10-17"
+          Statement:
+            - Effect: Allow
+              Action:
+                - xray:PutTraceSegments
+                - xray:PutTelemetryRecords
+                - cloudwatch:PutMetricData
+                - logs:PutLogEvents
+                - logs:CreateLogGroup
+                - logs:CreateLogStream
+              Resource: "*"
+
+        # Application Signals specific permissions
+        - Version: "2012-10-17"
+          Statement:
+            - Effect: Allow
+              Action:
+                - application-signals:PutServiceLevelObjective
+                - application-signals:GetServiceLevelObjective
+                - application-signals:ListServiceLevelObjectives
+              Resource: "*"
+
+  # CloudWatch Log Group with retention
+  PetHistoryLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Sub "/aws/lambda/${PetHistoryFunction}"
+      RetentionInDays: 14
+
+  # CloudWatch Alarms for monitoring
+  PetHistoryErrorAlarm:
+    Type: AWS::CloudWatch::Alarm
+    Properties:
+      AlarmName: !Sub "${PetHistoryFunction}-Errors"
+      AlarmDescription: "Pet History Lambda function error rate is too high"
+      MetricName: Errors
+      Namespace: AWS/Lambda
+      Statistic: Sum
+      Period: 300
+      EvaluationPeriods: 2
+      Threshold: 5
+      ComparisonOperator: GreaterThanThreshold
+      Dimensions:
+        - Name: FunctionName
+          Value: !Ref PetHistoryFunction
+      TreatMissingData: notBreaching
+
+  PetHistoryDurationAlarm:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

By setting these alarms, you can proactively monitor DynamoDB&#39;s capacity utilization and be alerted if your application approaches or exceeds the allocated capacity, or if throttle events occur, enabling you to take corrective actions.